evidence/store.go

```diff
@@ -0,0 +1,189 @@
+}
+
+const (
+	baseKeyLookup   = "evidence-lookup"   // all evidence
+	baseKeyOutqueue = "evidence-outqueue" // not-yet broadcast
+	baseKeyPending  = "evidence-pending"  // broadcast but not committed
+)
+
+func keyLookup(evidence types.Evidence) []byte {
+	return keyLookupFromHeightAndHash(evidence.Height(), evidence.Hash())
+}
+
+// big endian padded hex
+func be(h int) string {
+	return fmt.Sprintf("%0.16X", h)
+}
+
+func keyLookupFromHeightAndHash(height int, hash []byte) []byte {
+	return _key("%s/%s/%X", baseKeyLookup, be(height), hash)
+}
+
+func keyOutqueue(evidence types.Evidence, priority int) []byte {
+	return _key("%s/%s/%s/%X", baseKeyOutqueue, be(priority), be(evidence.Height()), evidence.Hash())
+}
+
+func keyPending(evidence types.Evidence) []byte {
+	return _key("%s/%s/%X", baseKeyPending, be(evidence.Height()), evidence.Hash())
+}
+
+func _key(fmt_ string, o ...interface{}) []byte {
+	return []byte(fmt.Sprintf(fmt_, o...))
+}
+
+// EvidenceStore is a store of all the evidence we've seen, including
+// evidence that has been committed, evidence that has been verified but not broadcast,
+// and evidence that has been broadcast but not yet committed.
+type EvidenceStore struct {
+	db dbm.DB
+}
+
+func NewEvidenceStore(db dbm.DB) *EvidenceStore {
+	return &EvidenceStore{
+		db: db,
+	}
+}
+
+// PriorityEvidence returns the evidence from the outqueue, sorted by highest priority.
+func (store *EvidenceStore) PriorityEvidence() (evidence []types.Evidence) {
+	// reverse the order so highest priority is first
+	l := store.ListEvidence(baseKeyOutqueue)
+	l2 := make([]types.Evidence, len(l))
+	for i := range l {
+		l2[i] = l[len(l)-1-i]
+	}
+	return l2
+}
+
+// PendingEvidence returns all known uncommitted evidence.
+func (store *EvidenceStore) PendingEvidence() (evidence []types.Evidence) {
+	return store.ListEvidence(baseKeyPending)
+}
+
+// ListEvidence lists the evidence for the given prefix key.
+// It is wrapped by PriorityEvidence and PendingEvidence for convenience.
+func (store *EvidenceStore) ListEvidence(prefixKey string) (evidence []types.Evidence) {
+	iter := store.db.IteratorPrefix([]byte(prefixKey))
+	for iter.Next() {
+		val := iter.Value()
+
+		var ei EvidenceInfo
+		wire.ReadBinaryBytes(val, &ei)
+		evidence = append(evidence, ei.Evidence)
+	}
+	return evidence
+}
+
+// GetEvidence fetches the evidence with the given height and hash.
+func (store *EvidenceStore) GetEvidence(height int, hash []byte) *EvidenceInfo {
+	key := keyLookupFromHeightAndHash(height, hash)
+	val := store.db.Get(key)
+
+	if len(val) == 0 {
+		return nil
+	}
+	var ei EvidenceInfo
+	wire.ReadBinaryBytes(val, &ei)
+	return &ei
+}
+
+// AddNewEvidence adds the given evidence to the database.
+// It returns false if the evidence is already stored.
+func (store *EvidenceStore) AddNewEvidence(evidence types.Evidence, priority int) bool {
+	// check if we already have seen it
+	ei_ := store.GetEvidence(evidence.Height(), evidence.Hash())
```

Comment:
What if a byzantine validator signed a million votes for the same height/round?

Reply:
I think for simple vote evidence, we could search for whether we have evidence for a validator at a given height, and not worry about rounds or blockhashes etc.  So if we already have evidence to slash a validator at a height, no need to keep collecting evidence.  What do you think?

Reply:
That sounds good. So we should change the lookup scheme to include the validator-address in the key so we can quickly check if theres already evidence for the validator at the height?